op-node/rollup/derive/channel_bank.go

```diff
@@ -143,73 +138,37 @@ func (ib *ChannelBank) Read() (data []byte, err error) {
 	return data, nil
 }
 
-// Step does the advancement for the channel bank.
-// Channel bank as the first non-pull stage does it's own progress maintentance.
-// When closed, it checks against the previous origin to determine if to open itself
-func (ib *ChannelBank) Step(ctx context.Context, _ Progress) error {
-	// Open ourselves
-	// This is ok to do b/c we would not have yielded control to the lower stages
-	// of the pipeline without being completely done reading from L1.
-	if ib.progress.Closed {
-		if ib.progress.Origin != ib.prev.Origin() {
-			ib.progress.Closed = false
-			ib.progress.Origin = ib.prev.Origin()
-			return nil
-		}
-	}
-
-	skipIngest := ib.next.Progress().Origin.Number > ib.progress.Origin.Number
-	outOfData := false
+// NextData pulls the next piece of data from the channel bank.
+// Note that it attempts to pull data out of the channel bank prior to
+// loading data in (unlike most other stages). This is to ensure maintain
+// consistency around channel bank pruning which depends upon the order
+// of operations.
+func (ib *ChannelBank) NextData(ctx context.Context) ([]byte, error) {
 
-	if data, err := ib.prev.NextData(ctx); err == io.EOF {
-		outOfData = true
+	// Do the read from the channel bank first
+	data, err := ib.Read()
+	if err == io.EOF {
+		// continue - We will attempt to load data into the channel bank
```

Comment:
maybe the below code can be nested here? I'm not sure what's best; flat code or easier coverage checking.

Reply:
it can be. Nice catch. I might do this cleanup later, but it's a really good callout